vercel/turbo
# feat: enable new UI

### Description

This PR adds an `App` widget and hooks it up to the task graph execution. The use of the TUI is controlled by the `TURBO_EXPERIMENTAL_UI` env var.

Note that the TUI is experimental so this current iteration does not attempt to tackle the following issues in this PR:
 - Proper integration with our `tracing` setup. Writing to stderr (or stdout) naively will cause us to overwrite the TUI as the cursor is probably not in the correct position. Currently the best way to run with increased verbosity is to pipe stderr to a file e.g. `TURBO_EXPERIMENTAL_UI=true turbo build -vv 2> /tmp/logs.txt`
 - "Proper" signal handling. Since we have [ISIG](https://man7.org/linux/man-pages/man3/termios.3.html) flagged off we are in charge of generating a SIGINT when `Ctrl-C` is pressed. We currently stop rendering the TUI and shut down on `Ctrl-C`, but we should do a [raise](https://man7.org/linux/man-pages/man3/raise.3.html)/Windows equivalent. https://github.com/vercel/turbo/pull/7737
 - Persistent configuration of enabling the TUI i.e. `turbo.json` extension
 - Proper interaction with existing flags that modify output e.g. `--log-order`/`--output-logs`
 - Responding to terminal resizing events
 - Final TUI state, already there are things I want to change e.g. have task render red on failure or rainbow on cache hit
 - Interactive tasks: #7735

Note for reviewers: Each commit is pretty self contained and should be reviewed on it's own.

### Testing Instructions

Primary thing that should be tested is that when `TURBO_EXPERIMENTAL_UI` is not set, there aren't any behavior changes.
The primary place where behavior could change is us changing reading task input to no longer be line based. This should be covered by #7728, but it is still worth manually checking since our prysk tests do not cover this codepath.

To test the TUI set `TURBO_EXPERIMENTAL_UI=true` and run `turbo`.


https://github.com/vercel/turbo/assets/4131117/9097dc07-c76a-40af-9a81-26fcb1c18a19
